src/main/java/org/opensearch/neuralsearch/processor/rerank/RerankProcessor.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.processor.rerank;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.neuralsearch.processor.rerank.context.ContextSourceFetcher;
+import org.opensearch.search.pipeline.PipelineProcessingContext;
+import org.opensearch.search.pipeline.SearchResponseProcessor;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+/**
+ * Abstract base class for reranking processors
+ */
+@AllArgsConstructor
+public abstract class RerankProcessor implements SearchResponseProcessor {
+
+    public static final String TYPE = "rerank";
+
+    protected final RerankType subType;
+    @Getter
+    private final String description;
+    @Getter
+    private final String tag;
+    @Getter
+    private final boolean ignoreFailure;
+    protected List<ContextSourceFetcher> contextSourceFetchers;
+
+    /**
+     * Generate the information that this processor needs in order to rerank.
+     * Concurrently hit all contextSourceFetchers
+     * @param searchRequest the search query
+     * @param searchResponse the search results, in case they're relevant
+     * @param listener be async
+     */
+    public void generateRerankingContext(
+        SearchRequest searchRequest,
+        SearchResponse searchResponse,
+        ActionListener<Map<String, Object>> listener
```

Comment:
ack. can I guarantee that the search response is immutable?

Reply:
I don't think it can be guaranteed, and also in some case we need to mutate the object, it's more like a basic safeguard from re-assigned the ref to a totally different object.